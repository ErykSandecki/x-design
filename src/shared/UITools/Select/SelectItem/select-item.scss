@import '../../../../styles/typography';

.SelectItem {
  @include theme-selector('blue-2', background-color, '&:hover');

  align-items: center;
  border-radius: 5px;
  column-gap: 10px;
  cursor: pointer;
  display: flex;
  font-family: Rubik, sans-serif;
  font-size: $--xd-typography-font-size-small;
  height: 24px;
  justify-content: left;
  line-height: $--xd-typography-line-height-small;
  padding: 0 5px;

  & * {
    pointer-events: none;
  }

  &__check-icon {
    flex-shrink: 0;
    margin-right: 2.5px;
    opacity: 0;

    &--selected {
      opacity: 1;
    }
  }
}
