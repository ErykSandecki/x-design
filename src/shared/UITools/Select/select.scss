.Select {
  &--disabled {
    cursor: not-allowed;
    pointer-events: none;
  }

  & * {
    cursor: pointer;
  }
}
