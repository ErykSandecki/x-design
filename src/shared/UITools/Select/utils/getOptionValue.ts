import { MouseEvent } from 'react';

export const getOptionValue = (event: MouseEvent<HTMLElement>): string | undefined => {
  const target = event.target as HTMLElement;

  if (target.tagName === 'LI') {
    return target.getAttribute('data-value') as string;
  }

  return undefined;
};
