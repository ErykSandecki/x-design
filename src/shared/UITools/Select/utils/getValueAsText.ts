export const getValueAsText = (isMixed: boolean, t: TT, translationNameSpace: string, value: string): string => {
  if (translationNameSpace && value && !isMixed) {
    return t(`${translationNameSpace}.${value}`);
  }

  return value;
};
