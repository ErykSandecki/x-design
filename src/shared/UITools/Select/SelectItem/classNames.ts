export const className = 'SelectItem';

// utils
import { composeClassNames } from 'utils';

export const classNames = composeClassNames(
  className,
  [className, 'disabled', 'selected'] as const,
  ['checkIcon', 'selected'] as const,
);
