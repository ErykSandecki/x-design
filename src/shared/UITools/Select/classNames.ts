// utils
import { composeClassNames } from 'utils';

export const className = 'Select';

export const classNames = composeClassNames(
  className,
  [className, 'disabled', 'fullWidth'] as const,
  ['input'] as const,
);
