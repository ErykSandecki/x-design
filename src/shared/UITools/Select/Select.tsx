import { FC, MouseEvent, ReactElement, useRef } from 'react';
import { useTranslation } from 'react-i18next';

// components
import Box from '../../UI/Box/Box';
import Icon from '../../UI/Icon/Icon';
import SelectOptions from './SelectOptions/SelectOptions';
import TextField, { TTextFieldProps } from '../TextField/TextField';

// hooks
import { useSelectEvents } from './hooks/useSelectEvents';
import { useTheme } from 'hooks';

// others
import { className, classNames } from './classNames';

// styles
import styles from './select.scss';

// types
import { E2EAttribute } from 'types';
import { TextFieldVariant } from '../TextField/enums';

// utils
import { getValueAsText } from './utils/getValueAsText';

export type TSelectProps = Omit<TTextFieldProps, 'endAdorment' | 'onChange'> & {
  children?: ReactElement | Array<ReactElement>;
  disabled?: boolean;
  enableTyping?: boolean;
  idContainerOptions?: string;
  isMixed?: boolean;
  onChange: TFunc<[string]>;
  onMouseEnterSelect?: TFunc<[MouseEvent]>;
  onMouseEnterOptions?: TFunc<[string]>;
  onMouseLeaveSelect?: TFunc<[MouseEvent]>;
  onMouseLeaveOptions?: TFunc<[string]>;
  translationNameSpace?: string;
  value: string;
};

export const Select: FC<TSelectProps> = ({
  children,
  disabled = false,
  e2eValue = '',
  enableTyping = false,
  idContainer,
  idContainerOptions,
  isMixed = false,
  onChange,
  onMouseEnterSelect,
  onMouseEnterOptions,
  onMouseLeaveSelect,
  onMouseLeaveOptions,
  translationNameSpace = '',
  variant = TextFieldVariant.outlined,
  value,
  ...restProps
}) => {
  const targetValue = isMixed ? 'Mixed' : value;
  const optionsRef = useRef<HTMLDivElement>(null);
  const selectRef = useRef(null);
  const wrapperRef = useRef(null);
  const { t } = useTranslation();
  const { classNamesWithTheme, cx } = useTheme(classNames, styles);
  const { onClickOption, onClickSelect, selected } = useSelectEvents(idContainer, onChange, selectRef);

  return (
    <Box
      classes={{
        className: cx(classNamesWithTheme[className].name, [
          classNamesWithTheme[className].modificators.disabled,
          disabled,
        ]),
      }}
      e2eAttribute={E2EAttribute.select}
      e2eValue={e2eValue}
      onClick={onClickSelect}
      onMouseEnter={onMouseEnterSelect}
      onMouseLeave={onMouseLeaveSelect}
      ref={selectRef}
    >
      <TextField
        disabled={disabled}
        endAdorment={<Icon height={5} name="ChevronDown" style={{ marginRight: '5px' }} width={8} />}
        readOnly={!enableTyping}
        value={getValueAsText(isMixed, t, translationNameSpace, targetValue)}
        variant={variant}
        wrapperRef={wrapperRef}
        {...restProps}
      />
      <SelectOptions
        e2eValue={e2eValue}
        idContainer={idContainerOptions}
        onClick={onClickOption}
        onMouseEnterOptions={onMouseEnterOptions}
        onMouseLeaveOptions={onMouseLeaveOptions}
        ref={optionsRef}
        selected={selected}
        value={targetValue}
        wrapperRef={wrapperRef}
      >
        {children}
      </SelectOptions>
    </Box>
  );
};

export default Select;
