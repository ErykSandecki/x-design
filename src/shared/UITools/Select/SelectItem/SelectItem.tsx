import { FC, ReactNode, useRef } from 'react';

// components
import Box from '../../../UI/Box/Box';
import E2EDataAttribute from 'shared/E2EDataAttributes/E2EDataAttribute';
import Icon from '../../../UI/Icon/Icon';

// hooks
import { useTheme } from 'hooks';

// others
import { className, classNames } from './classNames';

// styles
import styles from './select-item.scss';

// types
import { E2EAttribute } from 'types';

export type TSelectItemProps = {
  children: ReactNode;
  disabled?: boolean;
  index?: number;
  selectedValue?: string | Array<string>;
  value: string;
};

export const SelectItem: FC<TSelectItemProps> = ({ children, disabled = false, index = -1, selectedValue, value }) => {
  const ref = useRef<HTMLLIElement>(null);
  const isSelected = selectedValue === value;
  const { classNamesWithTheme, cx } = useTheme(classNames, styles);

  return (
    <E2EDataAttribute type={E2EAttribute.selectItem} value={index}>
      <Box
        classes={{
          className: cx(
            classNamesWithTheme[className].name,
            [classNamesWithTheme[className].modificators.disabled, disabled],
            [classNamesWithTheme[className].modificators.selected, isSelected],
          ),
        }}
        component="li"
        data-value={value}
        ref={ref}
        role="option"
        tabIndex={0}
      >
        <Icon
          classes={{
            className: cx(classNamesWithTheme.checkIcon.name, [
              classNamesWithTheme.checkIcon.modificators.selected,
              isSelected,
            ]),
          }}
          height={12}
          name="Check"
          width={12}
        />
        {children}
      </Box>
    </E2EDataAttribute>
  );
};

export default SelectItem;
