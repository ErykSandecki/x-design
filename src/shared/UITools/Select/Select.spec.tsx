import { fireEvent } from '@testing-library/react';

// components
import Select from './Select';
import SelectItem from './SelectItem/SelectItem';

// types
import { E2EAttribute, HTMLContainerId } from 'types';

// utils
import { customRender, getByE2EAttribute } from 'test';
import { createHtmlElement } from 'utils';

const dropdownContainer = createHtmlElement('div', { id: HTMLContainerId.dropdown });
const mockCallBack = jest.fn();

describe('Select snapshots', () => {
  beforeAll(() => {
    // mock
    document.body.appendChild(dropdownContainer);
  });

  it('should render Select', () => {
    // before
    const { asFragment } = customRender(
      <Select onChange={mockCallBack} value="0">
        <SelectItem value="0">0</SelectItem>
        <SelectItem value="1">1</SelectItem>
        <SelectItem value="2">2</SelectItem>
      </Select>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render when mixed', () => {
    // before
    const { asFragment } = customRender(
      <Select isMixed onChange={mockCallBack} value="0">
        <SelectItem value="0">0</SelectItem>
        <SelectItem value="1">1</SelectItem>
        <SelectItem value="2">2</SelectItem>
      </Select>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with no options', () => {
    // before
    const { asFragment } = customRender(<Select onChange={mockCallBack} value="0" />);

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with one option', () => {
    // before
    const { asFragment } = customRender(
      <Select onChange={mockCallBack} value="0">
        <SelectItem value="0">0</SelectItem>
      </Select>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Select behaviors', () => {
  beforeAll(() => {
    // mock
    document.body.appendChild(dropdownContainer);
  });

  it('should change value', () => {
    // before
    const { container } = customRender(
      <Select onChange={mockCallBack} value="0">
        <SelectItem value="0">0</SelectItem>
        <SelectItem value="1">1</SelectItem>
        <SelectItem value="2">2</SelectItem>
      </Select>,
    );

    // find
    const select = getByE2EAttribute(container, E2EAttribute.select);
    const selectOptions = getByE2EAttribute(select, E2EAttribute.selectOptions);
    const selectItem = getByE2EAttribute(selectOptions, E2EAttribute.selectItem, 2);

    // action
    fireEvent.click(select);
    fireEvent.click(selectItem);

    // result
    expect(mockCallBack.mock.calls[0][0]).toBe('2');
  });
});
