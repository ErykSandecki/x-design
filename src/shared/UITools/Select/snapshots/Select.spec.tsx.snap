// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select snapshots should render Select 1`] = `
<DocumentFragment>
  <div
    class="css-1jnxvbp Select Select--light"
    data-test-select=""
  >
    <div
      class="css-1jnxvbp TextField TextField--light"
      data-test-text-field=""
      data-test-tooltip=""
    >
      <div
        class="css-1jnxvbp TextFieldLabel TextFieldLabel--light TextFieldLabel--outlined TextFieldLabel--outlined--light"
        data-test-box=""
      >
        <input
          class="TextFieldLabel__input TextFieldLabel__input--light"
          data-test-text-field-input=""
          maxlength="6"
          readonly=""
          value="0"
        />
        <div
          class="css-1jnxvbp Icon Icon--light Icon--neutral1 Icon--neutral1--light"
          data-test-icon="chevron-down"
          height="5"
          name="ChevronDown"
          style="margin-right: 5px;"
          width="8"
        />
      </div>
    </div>
    <ul
      class="css-1jnxvbp SelectOptions SelectOptions--light"
      data-test-box=""
      data-test-select-options=""
      style="left: 0px; top: 24px; width: 0px;"
    >
      <li
        class="css-1jnxvbp SelectItem SelectItem--light SelectItem--selected SelectItem--selected--light"
        data-test-box=""
        data-test-select-item="0"
        data-value="0"
        role="option"
        tabindex="0"
      >
        <div
          class="css-1jnxvbp SelectItem__checkIcon SelectItem__checkIcon--light SelectItem__checkIcon--selected SelectItem__checkIcon--selected--light Icon Icon--light Icon--neutral1 Icon--neutral1--light"
          data-test-icon="check"
          height="12"
          name="Check"
          width="12"
        />
        0
      </li>
      <li
        class="css-1jnxvbp SelectItem SelectItem--light"
        data-test-box=""
        data-test-select-item="1"
        data-value="1"
        role="option"
        tabindex="0"
      >
        <div
          class="css-1jnxvbp SelectItem__checkIcon SelectItem__checkIcon--light Icon Icon--light Icon--neutral1 Icon--neutral1--light"
          data-test-icon="check"
          height="12"
          name="Check"
          width="12"
        />
        1
      </li>
      <li
        class="css-1jnxvbp SelectItem SelectItem--light"
        data-test-box=""
        data-test-select-item="2"
        data-value="2"
        role="option"
        tabindex="0"
      >
        <div
          class="css-1jnxvbp SelectItem__checkIcon SelectItem__checkIcon--light Icon Icon--light Icon--neutral1 Icon--neutral1--light"
          data-test-icon="check"
          height="12"
          name="Check"
          width="12"
        />
        2
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Select snapshots should render when mixed 1`] = `
<DocumentFragment>
  <div
    class="css-1jnxvbp Select Select--light"
    data-test-select=""
  >
    <div
      class="css-1jnxvbp TextField TextField--light"
      data-test-text-field=""
      data-test-tooltip=""
    >
      <div
        class="css-1jnxvbp TextFieldLabel TextFieldLabel--light TextFieldLabel--outlined TextFieldLabel--outlined--light"
        data-test-box=""
      >
        <input
          class="TextFieldLabel__input TextFieldLabel__input--light"
          data-test-text-field-input=""
          maxlength="6"
          readonly=""
          value="Mixed"
        />
        <div
          class="css-1jnxvbp Icon Icon--light Icon--neutral1 Icon--neutral1--light"
          data-test-icon="chevron-down"
          height="5"
          name="ChevronDown"
          style="margin-right: 5px;"
          width="8"
        />
      </div>
    </div>
    <ul
      class="css-1jnxvbp SelectOptions SelectOptions--light"
      data-test-box=""
      data-test-select-options=""
      style="left: 0px; top: 24px; width: 0px;"
    >
      <li
        class="css-1jnxvbp SelectItem SelectItem--light"
        data-test-box=""
        data-test-select-item="0"
        data-value="0"
        role="option"
        tabindex="0"
      >
        <div
          class="css-1jnxvbp SelectItem__checkIcon SelectItem__checkIcon--light Icon Icon--light Icon--neutral1 Icon--neutral1--light"
          data-test-icon="check"
          height="12"
          name="Check"
          width="12"
        />
        0
      </li>
      <li
        class="css-1jnxvbp SelectItem SelectItem--light"
        data-test-box=""
        data-test-select-item="1"
        data-value="1"
        role="option"
        tabindex="0"
      >
        <div
          class="css-1jnxvbp SelectItem__checkIcon SelectItem__checkIcon--light Icon Icon--light Icon--neutral1 Icon--neutral1--light"
          data-test-icon="check"
          height="12"
          name="Check"
          width="12"
        />
        1
      </li>
      <li
        class="css-1jnxvbp SelectItem SelectItem--light"
        data-test-box=""
        data-test-select-item="2"
        data-value="2"
        role="option"
        tabindex="0"
      >
        <div
          class="css-1jnxvbp SelectItem__checkIcon SelectItem__checkIcon--light Icon Icon--light Icon--neutral1 Icon--neutral1--light"
          data-test-icon="check"
          height="12"
          name="Check"
          width="12"
        />
        2
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Select snapshots should render with no options 1`] = `
<DocumentFragment>
  <div
    class="css-1jnxvbp Select Select--light"
    data-test-select=""
  >
    <div
      class="css-1jnxvbp TextField TextField--light"
      data-test-text-field=""
      data-test-tooltip=""
    >
      <div
        class="css-1jnxvbp TextFieldLabel TextFieldLabel--light TextFieldLabel--outlined TextFieldLabel--outlined--light"
        data-test-box=""
      >
        <input
          class="TextFieldLabel__input TextFieldLabel__input--light"
          data-test-text-field-input=""
          maxlength="6"
          readonly=""
          value="0"
        />
        <div
          class="css-1jnxvbp Icon Icon--light Icon--neutral1 Icon--neutral1--light"
          data-test-icon="chevron-down"
          height="5"
          name="ChevronDown"
          style="margin-right: 5px;"
          width="8"
        />
      </div>
    </div>
    <ul
      class="css-1jnxvbp SelectOptions SelectOptions--light"
      data-test-box=""
      data-test-select-options=""
      style="left: 0px; top: 24px; width: 0px;"
    >
      <li
        class="css-1jnxvbp SelectItem SelectItem--light SelectItem--disabled SelectItem--disabled--light"
        data-test-box=""
        data-test-select-item="-1"
        data-value=""
        role="option"
        tabindex="0"
      >
        <div
          class="css-1jnxvbp SelectItem__checkIcon SelectItem__checkIcon--light Icon Icon--light Icon--neutral1 Icon--neutral1--light"
          data-test-icon="check"
          height="12"
          name="Check"
          width="12"
        />
        No options
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Select snapshots should render with one option 1`] = `
<DocumentFragment>
  <div
    class="css-1jnxvbp Select Select--light"
    data-test-select=""
  >
    <div
      class="css-1jnxvbp TextField TextField--light"
      data-test-text-field=""
      data-test-tooltip=""
    >
      <div
        class="css-1jnxvbp TextFieldLabel TextFieldLabel--light TextFieldLabel--outlined TextFieldLabel--outlined--light"
        data-test-box=""
      >
        <input
          class="TextFieldLabel__input TextFieldLabel__input--light"
          data-test-text-field-input=""
          maxlength="6"
          readonly=""
          value="0"
        />
        <div
          class="css-1jnxvbp Icon Icon--light Icon--neutral1 Icon--neutral1--light"
          data-test-icon="chevron-down"
          height="5"
          name="ChevronDown"
          style="margin-right: 5px;"
          width="8"
        />
      </div>
    </div>
    <ul
      class="css-1jnxvbp SelectOptions SelectOptions--light"
      data-test-box=""
      data-test-select-options=""
      style="left: 0px; top: 24px; width: 0px;"
    >
      <li
        class="css-1jnxvbp SelectItem SelectItem--light SelectItem--selected SelectItem--selected--light"
        data-test-box=""
        data-test-select-item="0"
        data-value="0"
        role="option"
        tabindex="0"
      >
        <div
          class="css-1jnxvbp SelectItem__checkIcon SelectItem__checkIcon--light SelectItem__checkIcon--selected SelectItem__checkIcon--selected--light Icon Icon--light Icon--neutral1 Icon--neutral1--light"
          data-test-icon="check"
          height="12"
          name="Check"
          width="12"
        />
        0
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
